### PR TITLE
[SPARK-12543] [SPARK-4226] [SQL] Subquery in expression

### PR DESCRIPTION
**Pull request by davies**
This PR brings subquery for expression (use subquery as an expression inside SELECT/WHERE/HAVING), for example:

```
  select (select 1) as a
  select * from t where a = (select max(b) from t2)
  select * from t where exists (select * from t2 where t2.b = t.b)
  select max(a) as ma from t having ma = (select max(b) from t2)
```

A subquery could be uncorrelated or correlated, it could be scalar subquery (returns single row and single column) or not (returns multiple rows, for EXISTS or IN).

A correlated subquery or uncorrelated subquery that returns multiple rows will be rewritten as JOIN. Scalar subquery will be executed separately, result will be filled into the expression.

Restrictions:
1. EXISTS/ NOT EXISTS/IN/NOT IN can only be used as top level condition of WHERE/HAVING, can't be used in OR or others.
2. a NOT IN subquery is not supported, if subquery is nullable, which requires a null-aware semi join.
3. Correlated subquery can only be used inside WHERE/HAVING, uncorrelated scalar subquery could be used anywhere.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/SemiJoinSuite.scala [hunk 1]
```diff
@@ -108,9 +109,26 @@ class SemiJoinSuite extends SparkPlanTest with SharedSQLContext {
...
+    "anti join test",
+    left,
+    right,
+    condition,
+    LeftAnti,
+    Seq(
+      (1, 2.0),
+      (1, 2.0),
+      (3, 3.0),
+      (null, null),
```
Comment (chenghao-intel) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/SemiJoinSuite.scala]: This is actually what I am struggling, if the join key is `null`, then will we let's go into the anti join result?

**Comment by davies** [comment]
@chenghao-intel Since you also worked on this topic, could you take a look a this?

It's good to turn on those Hive compatibility tests, but I'm stucked by how to generated the golden results.  I tried to pull in yours in #9055, but unfortunately some of those queries are not in good state (missing spaces?).

**Comment by chenghao-intel** [comment]
Yes, I can build a PR against your branch. :), but for `anti join`, I have another PR #10563, some of my concerns are being discussed, one key issue is about the null join key, I think we probably need to filter out the `null` in `predicate` expression, and it will be great if we can figure out that issue first.

**Comment by davies** [comment]
@chenghao-intel This PR does not have a null-aware anti join, it will fail some query when analyzing, we could add that later.

**Comment by davies** [comment]
@hvanhovell Can you help to look at this one? I'd like to split this out as small PRs, hopefully we can merge part of them into 2.0.

**Comment by hvanhovell** [comment]
@davies sure, no problem.

**Comment by rxin** [comment]
Can we close this first, and create a new one when we get to correlated subqueries?

**Comment by davies** [comment]
I'd like to keep this open so I can easily find this branch.

**Comment by imperio-wxm** [comment]
Hi,I encountered a similar problem.(spark:1.5.2)

Subquery like this:
```sql
SELECT * FROM his_data_zadd WHERE value=(SELECT MAX(his_t.value) FROM his_data_zadd AS his_t)
```

Error code:
```sql
py4j.protocol.Py4JJavaError: An error occurred while calling o32.sql.
: java.lang.RuntimeException: [1.49] failure: ``)'' expected but identifier MAX found

SELECT * FROM his_data_zadd WHERE value=(SELECT MAX(his_t.value) FROM his_data_zadd AS his_t)
                                                ^
```

How should I write correctly subquery?

**Comment by davies** [comment]
Spark 1.5 does not support subquery.

**Comment by imperio-wxm** [comment]
Thanks.

**Comment by kamalcoursera** [comment]
Hi Davies,

Could you please shed more light on the status of correlated but non-scalar subquery in Spark 2.0 release. Appreciate if you can summarize any other restrictions, if any.


**Query:**

Select 
runon as runon,  
case 
 when key in (Select key from sqltesttable where group = 'vowels') then 'vowels'
  else 'consonants'
end as group,
key as key,
someint as someint  
from sqltesttable;

**Error:**

Error in SQL statement: AnalysisException: Predicate sub-queries can only be used in a Filter: Project [runon#4031 AS runon#4026,CASE WHEN predicate-subquery#4027 [(key#4033 = key#4037)] THEN vowels ELSE consonants END AS group#4028,key#4033 AS key#4029,someint#4034 AS someint#4030]
:  +- SubqueryAlias predicate-subquery#4027 [(key#4033 = key#4037)]
:     +- Project [key#4037]
:        +- Filter (group#4036 = vowels)
:           +- MetastoreRelation default, sqltesttable, None
+- MetastoreRelation default, sqltesttable, None
;

**Comment by davies** [comment]
predicate subquery (IN, EXISTS) in SELECT is not supported in 2.0, only supported in WHERE/HAVING.

**Comment by kamalcoursera** [comment]
Thank you ! Any alternative options to use instead of predicate subquery ? 
Is it in plan for amendment in 2.0?

Select 
runon as runon,
case 
when key in (Select key from sqltesttable where group = 'vowels') then 'vowels'
else 'consonants'
end as group,
key as key,
someint as someint
from sqltesttable;

**Comment by hvanhovell** [comment]
@kamalcoursera you could use a predicate scalar subquery here, i.e.:
```sql
select   runon as runon
         case
          when (select max(true) from sqltesttable b where b.key = a.key and group = 'vowels') then 'vowels'
          else 'consonants'
         end as group,
         key as key,
         someint as someint
from     sqltesttable a;
```

**Pull request closed without merging**